rfcs/react-components/convergence/event-handlers-event-type.md

```diff
@@ -0,0 +1,125 @@
+
+#### Solution 1 - cast the new event to the existing type
+
+```ts
+// Within our codebase:
+export type OpenPopoverEvents = KeyboardEvent | MouseEvent;
+
+const usePopover = () => {
+  const handleScroll = (e: Event) => {
+    onOpenChange(e as OpenPopoverEvents, data);
+  };
+};
+
+// In consumer implementations:
+const onOpenChange = (e: MouseEvent | KeyboardEvent) => {
+  console.log(e.shiftKey); // 💣 This would fail since shiftKey doesn't exist on 'Event'
+};
+const props: Props = { onOpenChange };
+```
+
+- Pros 👍: No modifications to types.
+- Cons 👎: Run-time type inaccuracies as illustrated in the code comment above.
+
+### 2. Solutions for both existing components and upcoming components:
+
+For new components without backward compatibility concerns, new type signatures can be directly adopted.
+
+#### Solution 2 - callbacks with generic event type:
+
+```ts
+// For an existing component
+export type PopoverProps = {
+  /**
+   * @deprecated Use onOpenChange2 instead.
+   */
+  onOpenChange?: (e: OpenPopoverEvents, data: OnOpenChangeData) => void;
+
+  onOpenChange2?: (e: OpenPopoverEvents | Event, data: OnOpenChangeData) => void;
+};
+
+// For a new component
+export type SomeProps = {
+  callback?: (e: MouseEvent | Event, data: SomeData) => void;
+};
+```
+
+- Pros 👍: Ensures correct typing and it can accommodate future, unforeseen event types.
+- Cons 👎:
+
+  1. Consumers will need to use type assertion or type predict to use a strongly-typed event:
+
+     ```ts
+     function isMouseEvent(e: Event): e is MouseEvent {
+       return e instanceof MouseEvent;
+     }
+     const onOpenChange = (e: Event | MouseEvent | TouchEvent) => {
+       console.log(e.clientX); // ❌ type error - Property 'clientX' does not exist on type 'Event'
+       if (isMouseEvent(e)) {
+         console.log(e.clientX); // ✅
+       }
+       if (e instanceof MouseEvent) {
+         console.log(e.clientX); // ✅
+       }
+     };
+     ```
+
+- Extra cons 👎 for existing components:
+  1. This is fragile - old `onOpenChange` won't be invoked when generic event happens (for example on scroll).
+  2. Code bloat - we have many callbacks, there will be a bunch of extra codes to keep both versions of callbacks working.
+
+Also perhaps there will be a case where generic event is inappropriate. For instance, if we can confidently assert that a callback only triggers on keyboard events. Although no such cases are identified yet, it remains a potential concern.
+
+#### Solution 3 - "Data Bag" Approach - strongly-typed events are passed in the data argument.
+
+Here, if the event could be a mouse or keyboard event, the callback receives two data fields (mouseEvent and keyboardEvent) which could be undefined.
+
+```ts
+export type OnOpenChangeData = { open: boolean; mouseEvent?: MouseEvent; keyboardEvent?: KeyboardEvent };
```

Comment:
interesting idea.  

I'm missing benefits of this approach in our and user land over explicit guards to check the event type for runtime/type-check time. 

for example user would still need to check if keyboardEvent or mouseEvent is present etc ( which is "the same" as using guard `isMouseEvenet(ev)` etc ).

for future (v9/10)  I'd rather like us adopting browser PointerEvents which are event mechanism for our modern and future era ( handles any device generically ) - thus introducing this approach would only add maintenance and API baggage .